### ISSUE TITLE
Partial trace doesn't work on state vector output

### ISSUE THREAD

**Issue by Strilanc** [quantumlib/Cirq#1632]
I realize this is not really how the method was intended to use, but it felt natural as I was writing it.

```
import cirq
c = cirq.Circuit.from_ops(cirq.H.on_each(*cirq.LineQubit.range(3)))
r = cirq.Simulator().simulate(c)
m = cirq.partial_trace(
            r.final_simulator_state.state_vector.reshape((2, 2, 2)), [0])
```

causes

```
operand has more dimensions than subscripts given in einstein sum, but no '...' ellipsis provided to broadcast the extra dimensions.
```

**Comment by peterse** [comment]
that reshape operation is incorrect. In general you have to promote the wavefunction to a density matrix with np.kron, then do the partial trace, then apply a "nearest kronecker product" factorization to get a wavefunction as the final form. Then this will give you a unique traced wf representation (mod global phase) IFF the circuit is completely unitary.

**Comment by dabacon** [comment]
This would be a nice feature, but if we add support for this it should not be called just "partial trace" which is a well defined mathematical operation (what is described is a composition of outer product and partial trace).

**Comment by peterse** [comment]
resolved in #1699 
for the record:
- `cirq.subwavefunction` attempts to factor a wavefunction into two pure states and return the pure wavefunction over the `kept_qubits` arg, or else fails
- `cirq.wavefunction_partial_trace_as_mixture` factors a wavefunction into a (possibly pure) mixture and returns a tuple representing that mixture.
